### ISSUE TITLE
Feature Request: Add User Authentication with different backends

### ISSUE THREAD

**Issue by xshadow** [cryptpad/cryptpad#174]
It would be great if the admin could be able to configure different user authentication backends, instead of relying on the build in database. 

Backends of interest:
- IMAP (by server and domain) 
- LDAP
- SMB
- FTP

**Comment by ansuz** [comment]
If you use a reverse proxy, you can put the authentication at that layer so the node process does not need any modification.

However that is implemented, nothing should hook into the login process implemented on the CryptPad client. The /login/ and /register/ pages derive secret keys from users' credentials, and as such, user confidentiality relies on these credentials not being known by the server.

**Comment by cjdelisle** [comment]
This type of thing is definitely possible but because the password is used to derive the root encryption keys, it is non-trivial to do it without effectively sharing all decryption keys with the LDAP server. However, if you are able to sponsor some development time, we could prioritize this on our roadmap. We're available for a discussion at sales@cryptpad.fr

**Comment by xshadow** [comment]
The general idea would be, to give the admin of the cryptpad the option to rely on a different back end , instead of using the built in database. 

For example IMAP Auth: 
If you have a mail server running , every user who can authenticate it self on the mail server does also have an account on the cryptpad service. Though users do not have to use different credentials and manage their pads. 

Other services for example Nextcloud provide this handy [feature](https://docs.nextcloud.com/server/9/admin_manual/configuration_user/user_auth_ftp_smb_imap.html) . I will write you an email. 

Due to the fact that this feature would expose another attack entry point for brute force attacks, it would be good to implement security features from the beginning. Just an idea, the client have to calculate a hash which is known by the server, every time a login is done.

**Comment by cjdelisle** [comment]
Copied from a private correspondence:

The matter of external login is a complex one, because we hash the
password using scrypt and use the result as a secret key for encrypting
the /drive/, which then has the keys to all of one's pads. So if we made
CryptPad authenticate using LDAP or IMAP in the obvious way, the Zero
Knowledge aspect would be gone because the authentication server would
hold the key.

There might be some possible way to do double password or such thing
which allows logging in while still protecting the confidentiality of
the pads, but it seems non-trivial.

**Comment by rysiekpl** [comment]
This would be a very welcome feature indeed.

I would be fine with a clearly defined trade-off: zero knowledge, or external authentication (choose one). If somebody is setting external auth for something like Cryptpad, that's often in an organizational setting where the zero knowledge part might not be crucial.

Another way to do this would be to have the external auth service provide an additonal attribute accessible only in the browser (OpenID Connect makes this possible via JWT tokens; almost certain SAML does too) and used by browser-side JS the way password is used currently.

**Comment by ansuz** [comment]
I'm going to close this issue since we have no plans to implement the requested behaviour. If you're interested in sponsoring development costs I'll certainly reconsider. In that case, you can contact sales@cryptpad.fr

**Comment by rasos** [comment]
oAuth / OpenID Connect would be great to be able to offer a cryptpad instance it in larger organizations, who have tier own identity management system (such as keycloak).

**Comment by sbernhard** [comment]
Thats true, it would be great to have openid-connect and/or SAML authentication.

**Comment by StephanRichter** [comment]
I really would appreciate the OpenID Connection login. Also, I would be willing to pay some money – the question here is: how much would be appropriate?

**Comment by ansuz** [comment]
> I really would appreciate the OpenID Connection login. Also, I would be willing to pay some money – the question here is: how much would be appropriate?

I appreciate your willingness to compensate us for our time! That's unfortunately rare in open-source.

I would expect at least one week of full-time development to integrate OpenID into both the client and the server. Our standard rate for one day of consulting is 1000€, so assume a minimum of 5000€ for a first version and more if additional requirements are introduced.

If multiple organizations were willing to sponsor this development in coordination it would become much more affordable. Perhaps @xshadow, @rasos, @rysiekpl, or @sbernhard could help to split the costs?

**Comment by poVoq** [comment]
Not sure what came out of this:
https://nlnet.nl/project/CryptPad-Auth/

**Comment by timmwille** [comment]
> Not sure what came out of this: https://nlnet.nl/project/CryptPad-Auth/

thx for the link, I could not find further Info on this matter, could anyone answer what is the result/status?

**Comment by matacino** [comment]
> > Not sure what came out of this: https://nlnet.nl/project/CryptPad-Auth/
> 
> thx for the link, I could not find further Info on this matter, could anyone answer what is the result/status?

I stumbled upon: https://blog.cryptpad.org/2023/02/28/status-2023-02/

```
🔑 🪪 AUTH officially starts
We are happy to announce that we have finally signed off the CryptPad AUTH project …
```

**Comment by mxmilkiib** [comment]
Could this issue be reopened please given there is work being done on this but is not complete?

LDAP would be nice so a Cryptpad install I'm trying (for a small community group) can use the YunoHost turnkey private cloud distro SSO. (No idea how Nextcloud SSO works.)

**Comment by alexanderadam** [comment]
[This PR](https://github.com/cryptpad/cryptpad/pull/1320) looks very promising! :star_struck:

**Comment by bain3** [comment]
Hi! I'd just like to leave a note here that people need to install the SSO plugin:

https://github.com/cryptpad/sso

I've just spent a good hour diagnosing why `/api/config` was returning `sso: false`, even though I had filled in the SSO config file :)

**Comment by miharix** [comment]
Hi!

I didn't found clear current state info... 

Is LDAP supported or not ?
What other methods are available for connecting local instance to MS Active Directory other then SSO ?
When connected to AD is it still possible to add guest users ?

Cryptpad seems like professional looking solution for what we need but AD connection is must.
I will try my best to convince my boss to to donate(how much?) if the feature is not yet implemented.

**Comment by rfpg1** [comment]
I've the same question as @miharix 

Is LDAP already supported? Can't seem to find an answer

**Comment by ghost** [comment]
> Hi!
> 
> I didn't found clear current state info...
> 
> Is LDAP supported or not ? What other methods are available for connecting local instance to MS Active Directory other then SSO ? When connected to AD is it still possible to add guest users ?
> 
> Cryptpad seems like professional looking solution for what we need but AD connection is must. I will try my best to convince my boss to to donate(how much?) if the feature is not yet implemented.

-----

> I've the same question as [@miharix](https://github.com/miharix)
> 
> Is LDAP already supported? Can't seem to find an answer

-----

@miharix / @rfpg1 :: the linked repo of https://github.com/cryptpad/sso says:

> OIDC and SAML SSO connectors;

So no LDAP.
